### DISCUSSION TITLE
How to use sandbox together with JSON config?

### DISCUSSION THREAD

**Discussion by letmaik** [microsoft/CCF#3710] [Q&A]
I'm hitting https://github.com/microsoft/CCF/issues/2802 during testing with the sandbox and wanted to try increasing `max_msg_size`. The sandbox script doesn't seem to support this parameter, so I tried using a config file. I thought I may be able to augment whatever config file the sandbox generates itself but it looks like a full config is required. This includes specifying member keys etc., which the sandbox would normally have taken care of. It seems overly complicated to get this working, am I missing something?

**Comment by achamayou** [comment]
@letmaik the sandbox exists to allow quick experimentation with CCF apps, it is not a general purpose runtime environment. There is no plan to extend it further or allow arbitrary configuration to be passed, because:

1. it's not necessary for the purpose that the sandbox serves
2. it would be difficult to make sure that the infra is compatible with all possible configuration variants

**Reply by letmaik** [reply]
Hm, the page https://microsoft.github.io/CCF/main/build_apps/run_app.html says that the sandbox is useful for integration tests and load testing. Being able to influence parameters such that they are closer to a production system is useful I think. The alternative would be to not use the sandbox and create a custom sandbox-like or test system, which seems like a big leap.
Any reason that the sandbox can't merge a provided JSON config with its own generated JSON config?

**Reply by achamayou** [reply]
Yes, the reason is the infra code relies on some elements of the configuration being a certain way, eg. file paths. The sandbox is not a production environment or a generic way to run CCF nodes.

It's possible to tweak the config beyond what's supported but it can break easily, and the sandbox has no supported API.

**Comment by jumaffre** [comment]
I added a `--config-file` option to the generic infra to when doing the JSON configuration work (https://github.com/microsoft/CCF/blob/main/tests/infra/e2e_args.py#L341-L345) to support this. However, it's not tested in the CI so unlikely to work without some minor tweaks as we've changed the infra slightly since (e.g. support for multiple interfaces).
